pkg/storegateway/indexheader/stream_binary_reader.go

```diff
@@ -70,28 +70,31 @@ type StreamBinaryReader struct {
 
 // NewStreamBinaryReader loads or builds new index-header if not present on disk.
 func NewStreamBinaryReader(ctx context.Context, logger log.Logger, bkt objstore.BucketReader, dir string, id ulid.ULID, sparsePersistenceEnabled bool, postingOffsetsInMemSampling int, metrics *StreamBinaryReaderMetrics, cfg Config) (*StreamBinaryReader, error) {
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, logger, "indexheader.NewStreamBinaryReader")
+	defer spanLog.Finish()
+
 	binPath := filepath.Join(dir, id.String(), block.IndexHeaderFilename)
 	sparseHeadersPath := filepath.Join(dir, id.String(), block.SparseIndexHeaderFilename)
-	br, err := newFileStreamBinaryReader(ctx, binPath, id, sparseHeadersPath, sparsePersistenceEnabled, postingOffsetsInMemSampling, logger, metrics, cfg)
+	br, err := newFileStreamBinaryReader(binPath, id, sparseHeadersPath, sparsePersistenceEnabled, postingOffsetsInMemSampling, spanLog, metrics, cfg)
 	if err == nil {
 		return br, nil
 	}
 
-	level.Debug(logger).Log("msg", "failed to read index-header from disk; recreating", "path", binPath, "err", err)
+	level.Debug(spanLog).Log("msg", "failed to read index-header from disk; recreating", "path", binPath, "err", err)
 
 	start := time.Now()
 	if err := WriteBinary(ctx, bkt, id, binPath); err != nil {
 		return nil, fmt.Errorf("cannot write index header: %w", err)
 	}
 
-	level.Debug(logger).Log("msg", "built index-header file", "path", binPath, "elapsed", time.Since(start))
-	return newFileStreamBinaryReader(ctx, binPath, id, sparseHeadersPath, sparsePersistenceEnabled, postingOffsetsInMemSampling, logger, metrics, cfg)
+	level.Debug(spanLog).Log("msg", "built index-header file", "path", binPath, "elapsed", time.Since(start))
+	return newFileStreamBinaryReader(binPath, id, sparseHeadersPath, sparsePersistenceEnabled, postingOffsetsInMemSampling, spanLog, metrics, cfg)
```

Comment:
We're passing `spanLog` to `newFileStreamBinaryReader`, but span will be finished when `NewStreamBinaryReader` returns. Is this correct?

Reply:
The span will be finished yes but the stream reader doesn't store the logger or use it after the initialization done in `newFileStreamBinaryReader()`

Reply:
Sure, I can change the type.

Reply:
> The span will be finished yes but the stream reader doesn't store the logger or use it after the initialization done in `newFileStreamBinaryReader()`

Thanks 👍